### PR TITLE
Add DataTypeDomainBool

### PR DESCRIPTION
**Pull request by hczhcz**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):

Add "Bool" using UInt8 as its underlying type

Detailed description (optional):

```sql
CREATE TABLE test (a Bool) ENGINE = Memory;
INSERT INTO test VALUES ('false'), ('true');
SELECT * FROM test;
> false
> true
```

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Missing test.

File: dbms/src/DataTypes/DataTypeDomainBool.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            ostr.write(str_false, sizeof(str_false));
+    }
+
+    void deserializeText(IColumn & column, ReadBuffer & istr, const FormatSettings &) const override
+    {
+        ColumnUInt8 * col = typeid_cast<ColumnUInt8 *>(&column);
+
+        if (!col)
+            throw Exception(
+                String(getName()) + " domain can only deserialize columns of type UInt32."
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeDomainBool.cpp]: Error in comment.

File: dbms/src/DataTypes/DataTypeDomainBool.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                    + column.getName(),
+                ErrorCodes::ILLEGAL_COLUMN
+            );
+
+        char buffer[6] = {'\0'};
+        istr.read(buffer, sizeof(buffer) - 1);
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeDomainBool.cpp]: Extra data is eaten by this call.

Comment (alexey-milovidov) [reply dbms/src/DataTypes/DataTypeDomainBool.cpp]: Look at `void DataTypeNumberBase<T>::deserializeTextJSON` for example.

File: dbms/src/DataTypes/DataTypeDomainBool.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+        if (strcmp(buffer, str_true) == 0)
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeDomainBool.cpp]: Buffer overflow here.

File: dbms/src/DataTypes/DataTypeDomainBool.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            col->insert(1);
+        else if (strcmp(buffer, str_false) == 0)
+            col->insert(0);
+        else
+            throw Exception(
+                "Invalid boolean value.",
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeDomainBool.cpp]: Need to mention, what exactly value is invalid, in exception message.

File: dbms/src/DataTypes/DataTypeDomainBool.cpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                ErrorCodes::ILLEGAL_COLUMN
+            );
+
+        if (col->getData()[row_num])
+            ostr.write(str_true, sizeof(str_true));
+        else
+            ostr.write(str_false, sizeof(str_false));
+    }
+
+    void deserializeText(IColumn & column, ReadBuffer & istr, const FormatSettings &) const override
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeDomainBool.cpp]: 0 and 1 should be supported as well, for convenience.

File: dbms/src/DataTypes/DataTypeDomainBool.cpp [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        // the same as text deserialization
+        deserializeText(column, istr, settings);
+    }
+};
+
+}
+
+void registerDataTypeDomainBool(DataTypeFactory & factory)
+{
+    factory.registerDataTypeDomain("UInt8", std::make_unique<DataTypeDomainBool>());
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeDomainBool.cpp]: Mistake.

File: dbms/src/DataTypes/DataTypeDomainBool.cpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int CANNOT_PARSE_DOMAIN_VALUE_FROM_STRING;
+}
+
+namespace
+{
+
+class DataTypeDomainBool : public DataTypeDomainWithSimpleSerialization
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypeDomainBool.cpp]: `final`

For devirtualization in forwarded virtual calls like `serializeTextJSON`.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Some changes required.

**Comment by alexey-milovidov** [comment]
Now it looks Ok.

BTW, Should we support case insensitive values: `TRUE`, `True`, etc?
It depends on the origin of the motivation for this data type.

Also please make some research about Bool type in other DBMS.
Maybe we should peek synonims for this data type, like BIT, Boolean?
But if we do, it can be incompatible due to different representation.

**Comment by alexey-milovidov** [comment]
Should this type also represent a constraint
(what will happen if we write `INSERT INTO t SELECT 5 AS b`)

**Comment by alexey-milovidov** [comment]
Don't you think that predefined `Enum8('false' = 0, 'true' = 1)` as a synonim for Bool type can be easier and better solution?

**Comment by hczhcz** [comment]
@alexey-milovidov 

The motivation (and the implementation) is similar to IPv4/IPv6 domains in #3669 .

* Bool is useful when we match the type systems between ClickHouse and other systems like JSON, programming language bindings, etc.
* It is also good for documentation (in comparison to a raw UInt8)
* It is painful if we add a separate data type and make the type checking code more complex
* So, the approach in #3669 looks good: providing a "view" to an underlying type

> Should we support case insensitive values: TRUE, True, etc

For data exchanging, I would use one value.

> Should this type also represent a constraint?

To use 0 and 1 is more a "suggestion" considering things like `SELECT and(5, 6);` are legal now.

> Don't you think that predefined Enum8('false' = 0, 'true' = 1) as a synonim for Bool type can be easier and better solution?

The motivation should explain it. The reason I choose UInt8 is just because `SELECT toTypeName(1 = 1);` is UInt8.

**Comment by alexey-milovidov** [comment]
Ok.

**Comment by alexey-milovidov** [comment]
We have to check if `x = 'true'` works in SQL query.
BTW, does `x = '127.0.0.1'` works for IPv4 domain?

If not, it is the cause of major usability issues.

**Comment by orloffv** [comment]
is nullable supported?

**Comment by hczhcz** [comment]
@alexey-milovidov 

For now, both of them generate type errors.
We may use `select x;` for bool values and `select x = toIPv4('127.0.0.1')` for ipv4 values.

Two things might be useful:
1. Add a `toBool(x)` function to perform type conversions
2. Return bool type in comparison/logic functions

**Comment by hczhcz** [comment]
@orloffv `Nullable(Bool)` works.

**Comment by geldot** [comment]
In both MySQL and PostgreSQL, TRUE/FALSE are keywords, which makes sense independently of this PR. MySQL:

```
SELECT true, false, TRUE, FALSE, True, False;
-- 1 0 1 0 1 0
```

In MySQL it's a plain 1/0 bit, in PostgreSQL it's a special type similar to this patch with pretty formatting.

PostgreSQL accepts TRUE, FALSE, 1, 0, but also the strings 'true', 'false', 't', 'f', 'ON', 'OFF' and a bunch of other English-language strings.

I like "Bool" for the type name because it's consistent with the C++-style type naming scheme in Clickhouse.

Agree it'd be interesting to see how this plays with Bitmap, bit shifts, different input/output formats etc. a little more before landing the patch as there's no going back afterwards.

**Comment by alexey-milovidov** [comment]
This PR looks Ok, but I'm not ready to

> Return bool type in comparison/logic functions

because it will introduce more complications.
It means, that `Bool` type will be not the first class citizen.
But it still makes sense. I need to discuss this feature with my collegues a bit more, and I think we will merge.

**Comment by abyss7** [comment]
We decided to close PR and postpone this feature for now.

**Comment by alexey-milovidov** [comment]
We decided to continue :) #30684

**Pull request closed without merging**